src/directives.adoc

```diff
@@ -0,0 +1,308 @@
+)
+----
+
+More formally, here's the relevant portion of the <<sec:grammar,domain grammar>>:
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=encoding-directive]
+|===
+
+The directive has four sections: declare currently available modules to retain, declare
+additional modules to make available, define the new symbol table, define the new macro table.
+
+
+==== Retaining Available Modules
+
+An encoding directive defines a new encoding environment in terms of the current environment
+(that is, the encoding environment for the segment containing the directive).
+By default, the new environment starts with an empty set of available modules, and if any modules
+are to be reused by the new segment, they must be explicitly retained.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=retention]
+|===
+
+Before declaring new names, the directive can selectively `*retain*` available modules (that is,
+modules declared in the preceding directive.
+This is done either by using the keyword `*{asterisk}*` to copy
+all available modules from the current encoding environment into the new one, or by enumerating specific names to copy.
+
+
+==== Declaring Modules
+
+After possibly retaining modules from the current environment, the directive can make additional
+modules available, either loading them from the implementation’s catalog, or defining them inline.
+Either way, an entry is added (or updated) in the directive’s map of available modules.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=module-decl]
+|===
+
+The names of available modules can be remapped: if a name is reused, the earlier declaration is
+shadowed through the rest of the directive (including upcoming inline modules).
+
+
+===== Loading Shared Modules
+
+To make a shared module available, it must first be _loaded_, which gives the module a symbolic
+name that can be used to reference the module’s components.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=load]
+|===
+
+This works like an `import` struct in `$ion_symbol_table` in that it acquires an entity from the
+implementation’s catalog, though here there is no direct effect on the symbol table.
+The _catalog-name_, _catalog-version_, and _symbol-maxid_ arguments have the same meaning as the
+corresponding fields of an `imports` struct, but only the latter is optional.
+Resolving the name and version to a shared module is the same as for shared symbol tables, using
+the same algorithm for inexact match on the version.
+
+TIP: A primary design tenet of Ion 1.1 is to remain compatible with existing catalog APIs and
+services that vend shared symbol tables.
+Existing shared symbol tables are shared modules that export no macros.
+
+As suggested by its name, the _symbol-maxid_ argument only affects symbol allocation, not macros.
+Use of macros from a shared module requires exact match of the _shared-version_, and a module that
+was imported inexactly will trigger an error if its name appears within a local module or
+`*macro_table*`.
```

Comment:
This is in support of 1.0-style SSTs that have tunneled macro definitions, right? Allowing 1.0 consumers to keep using the same SST when they don't need the macros? If so, that makes sense. However, it does feel pretty complex.

Reply:
It primarily falls out from having module loading for symbols and for macros use the same syntax and "namespace" of available modules.  At the time of `load` we don't know how the module will be used, so as long as we support maxId behavior for symbols then the inexact match is possible.

I didn't think this was complex, just a constraint that's easily detected and messaged.  Can you explain what feels complicated here?